regression/cuda/test_conv2d.c

```diff
@@ -0,0 +1,283 @@
+}
+
+// Takes an MxN matrix A and a HxW filter, a padding P,
+// a vertical stride Sy and a horizontal Sx, and outputs
+// the result of a 2D convolution into B. B must be of size
+// output_dim(M, H, P, Sy) x (N, W, P, Sx)
+void conv2d(const float *A,
+            const int M,
+            const int N,
+            const float *filter,
+            const int H,
+            const int W,
+            const int P,
+            float *B,
+            const int Sy,
+            const int Sx)
+{
+        int result_h = output_dim(M, H, P, Sy);
+        int result_w = output_dim(N, W, P, Sx);
+        for(int by = 0; by < result_h; by++)
+                for(int bx = 0; bx < result_w; bx++)
+                {
+                        int window_y = by * Sy;
+                        int window_x = bx * Sx;
+                        
+                        float res = 0;
+                        for(int fy = 0; fy < H; fy++)
+                                for(int fx = 0; fx < W; fx++)
+                                {
+                                        int ay = window_y - P + fy;
+                                        int ax = window_x - P + fx;
+                                        float a = 0;
+                                        
+                                        if((0 <= ay && ay < M) &&
+                                           (0 <= ax && ax < N))
+                                                a = A[ay * N + ax];
+                                        res += filter[fy * W + fx] * a;
+                                 }
+                        B[by * result_w + bx] = res;
+                }
+}
+
+int kernel_conv2d(int argc, char **argv)
+{       
+        bsg_pr_test_info("Running CUDA Conv2D Kernel on a 2x2 tile group.\n\n");
+        char *elf, *test_name;
+        struct arguments_path args = { NULL, NULL };
+        argp_parse(&argp_path, argc, argv, 0, 0, &args);
+        elf = args.path;
+        test_name = args.name;
+
+        srand(time(0));
+        int rc;
+        hb_mc_device_t manycore, *mc = &manycore;
+        rc = hb_mc_device_init(mc, test_name, 0);
+        if(rc != HB_MC_SUCCESS)
+        {
+                bsg_pr_err("Failed to initialize device.\n");
+                return rc;
+        }
+
+        rc = hb_mc_device_program_init(mc, elf, ALLOC_NAME, 0);
+        if(rc != HB_MC_SUCCESS)
+        {
+                bsg_pr_err("Failed to initialize the program.\n");
+                return rc;
+        }
+
+        int M = generate_inclusive(MIN_N, MAX_N);
+        int N = generate_inclusive(MIN_N, MAX_N);
+        int H = generate_inclusive(MIN_F, MAX_F < N ? MAX_F : N);
+        int W = generate_inclusive(MIN_F, MAX_F < N ? MAX_F : N);
+        int P = generate_inclusive(MIN_P, MAX_P);
+        int Sx = generate_inclusive(MIN_S, MAX_S);
+        int Sy = generate_inclusive(MIN_S, MAX_S);
+        int Bx = output_dim(M, H, P, Sy);
+        int By = output_dim(N, W, P, Sx);
```

Comment:
In the code I linked previously I generate for doubles, and then convert to float/int/int16/int8